drivers/timer/silabs_sleeptimer_timer.c

```diff
@@ -0,0 +1,161 @@
+
+/* MAX_TIMEOUT_CYC expressed as ticks */
+static uint32_t g_max_timeout_ticks;
+
+/* Value of sleeptimer counter when the previous kernel tick was announced */
+static atomic_t g_last_count;
+
+/* Spinlock to sync between Compare ISR and update of Compare register */
+static struct k_spinlock g_lock;
+
+/* Set to true when timer is initialized */
+static bool g_init;
+
+static sl_sleeptimer_timer_handle_t sys_timer;
+
+static void sleeptimer_cb(sl_sleeptimer_timer_handle_t *handle, void *data)
+{
+	ARG_UNUSED(handle);
+	ARG_UNUSED(data);
+	uint32_t curr = sl_sleeptimer_get_tick_count();
+	uint32_t prev = atomic_get(&g_last_count);
+	uint32_t pending = curr - prev;
+
+	/* Number of unannounced ticks since the last announcement */
+	uint32_t unannounced = pending / g_cyc_per_tick;
+
+	atomic_set(&g_last_count, prev + unannounced * g_cyc_per_tick);
+
+	sys_clock_announce(unannounced);
+}
+
+void sys_clock_set_timeout(int32_t ticks, bool idle)
+{
+	ARG_UNUSED(idle);
+
+	if (!IS_ENABLED(CONFIG_TICKLESS_KERNEL)) {
+		return;
+	}
+
+	ticks = (ticks == K_TICKS_FOREVER) ? g_max_timeout_ticks : ticks;
+	ticks = CLAMP(ticks, 0, g_max_timeout_ticks);
+
+	k_spinlock_key_t key = k_spin_lock(&g_lock);
+
+	uint32_t curr = sl_sleeptimer_get_tick_count();
+	uint32_t prev = atomic_get(&g_last_count);
+	uint32_t pending = curr - prev;
+	uint32_t next = ticks * g_cyc_per_tick;
+
+	/* Next timeout is N ticks in the future, minus the current progress
+	 * towards the timeout. If we are behind, set the timeout to the first
+	 * possible upcoming tick.
+	 */
+	while (next < (pending + MIN_DELAY_CYC)) {
+		next += g_cyc_per_tick;
+	}
+	next -= pending;
+
+	sl_sleeptimer_restart_timer(&sys_timer, next, sleeptimer_cb, NULL, 0, 0);
+	k_spin_unlock(&g_lock, key);
+}
+
+uint32_t sys_clock_elapsed(void)
+{
+	if (!IS_ENABLED(CONFIG_TICKLESS_KERNEL) || !g_init) {
+		/* No unannounced ticks can have elapsed if not in tickless mode */
+		return 0;
+	} else {
+		return (sl_sleeptimer_get_tick_count() - atomic_get(&g_last_count)) /
+		       g_cyc_per_tick;
+	}
+}
+
+uint32_t sys_clock_cycle_get_32(void)
+{
+	return g_init ? sl_sleeptimer_get_tick_count() : 0;
+}
+
+static int sleeptimer_init(void)
+{
+	sl_status_t status = SL_STATUS_OK;
+
+	IRQ_CONNECT(DT_IRQ(DT_RTC, irq), DT_IRQ(DT_RTC, priority),
+		    CONCAT(DT_STRING_UPPER_TOKEN_BY_IDX(DT_RTC, interrupt_names, 0), _IRQHandler),
+		    0, 0);
+
+	sl_sleeptimer_init();
+
+	z_clock_hw_cycles_per_sec = sl_sleeptimer_get_timer_frequency();
+
+	BUILD_ASSERT(CONFIG_SYS_CLOCK_TICKS_PER_SEC > 0,
+		     "Invalid CONFIG_SYS_CLOCK_TICKS_PER_SEC value");
+
+	g_cyc_per_tick = z_clock_hw_cycles_per_sec / CONFIG_SYS_CLOCK_TICKS_PER_SEC;
+
+	__ASSERT(g_cyc_per_tick >= MIN_DELAY_CYC,
+		 "A tick of %u cycles is too short to be scheduled "
+		 "(min is %u). Config: SYS_CLOCK_TICKS_PER_SEC is "
+		 "%d and timer frequency is %u",
+		 g_cyc_per_tick, MIN_DELAY_CYC, CONFIG_SYS_CLOCK_TICKS_PER_SEC,
+		 z_clock_hw_cycles_per_sec);
```

Comment:
I guess it's technically not true, since the timer isn't restarted/reconfigured, so the N clock cycle delay of writing to LF peripheral registers doesn't apply. I don't see any harm in keeping the restriction for now, though.